llms/hf_llm/convert.py

```diff
@@ -0,0 +1,159 @@
+# Copyright © 2023 Apple Inc.
+
+import argparse
+import copy
+import glob
+import json
+from pathlib import Path
+
+import mlx.core as mx
+import mlx.nn as nn
+import transformers
+from huggingface_hub import snapshot_download
+from mlx.utils import tree_flatten, tree_map
+from models import Model, ModelArgs
+
+
+def fetch_from_hub(hf_path: str):
+    model_path = snapshot_download(
+        repo_id=hf_path,
+        allow_patterns=["*.json", "*.safetensors", "tokenizer.model"],
+    )
+    weight_files = glob.glob(f"{model_path}/*.safetensors")
+    if len(weight_files) == 0:
+        raise FileNotFoundError("No safetensors found in {}".format(model_path))
+
+    weights = {}
+    for wf in weight_files:
+        weights.update(mx.load(wf).items())
+
+    config = transformers.AutoConfig.from_pretrained(hf_path)
+    tokenizer = transformers.AutoTokenizer.from_pretrained(
+        hf_path,
+    )
+    return weights, config.to_dict(), tokenizer
+
+
+def quantize(weights, config, args):
+    quantized_config = copy.deepcopy(config)
+
+    # Load the model:
+    model = Model(ModelArgs.from_dict(config))
+    weights = tree_map(mx.array, weights)
+    model.load_weights(list(weights.items()))
+
+    # Quantize the model:
+    nn.QuantizedLinear.quantize_module(model, args.q_group_size, args.q_bits)
+
+    # Update the config:
+    quantized_config["quantization"] = {
+        "group_size": args.q_group_size,
+        "bits": args.q_bits,
+    }
+    quantized_weights = dict(tree_flatten(model.parameters()))
+
+    return quantized_weights, quantized_config
+
+
+def make_shards(weights: dict, max_file_size_gibibyte: int = 15):
+    max_file_size_bytes = max_file_size_gibibyte << 30
+    shards = []
+    shard, shard_size = {}, 0
+    for k, v in weights.items():
+        estimated_size = v.size * v.dtype.size
+        if shard_size + estimated_size > max_file_size_bytes:
+            shards.append(shard)
+            shard, shard_size = {}, 0
+        shard[k] = v
+        shard_size += estimated_size
+    shards.append(shard)
+    return shards
+
+
+def upload_to_hub(path: str, name: str):
+    from huggingface_hub import HfApi, logging
+
+    logging.set_verbosity_info()
+
+    api = HfApi()
+
+    repo_id = f"mlx-community/{name}"
+    api.create_repo(repo_id=repo_id, exist_ok=True)
+    api.upload_folder(
+        folder_path=path,
+        repo_id=repo_id,
+        repo_type="model",
+        multi_commits=True,
+        multi_commits_verbose=True,
+    )
```

Comment:
Updated snippet to create a model card that refers to the original model, reuses metadata and adds the `mlx` tag: https://gist.github.com/pcuenca/4f4839325e799183bc6b81c668a24161

(Sorry, can't submit inline as the preview gets confused with the backticks)